publisher/publisher.go

```diff
@@ -0,0 +1,288 @@
+		return
+	}
+	pub.CT = ctConfig
+	bundle, err := core.LoadCertBundle(ctConfig.BundleFilename)
+	if err != nil {
+		return
+	}
+	for _, cert := range bundle {
+		pub.CT.IssuerBundle = append(pub.CT.IssuerBundle, base64.StdEncoding.EncodeToString(cert.Raw))
+	}
+	ctBackoff, err := time.ParseDuration(ctConfig.SubmissionBackoffString)
+	if err != nil {
+		return
+	}
+	pub.CT.SubmissionBackoff = ctBackoff
+
+	return
+}
+
+func (pub *PublisherAuthorityImpl) submitToCTLog(serial string, jsonSubmission []byte, log LogDescription, client http.Client) error {
+	done := false
+	var retries int
+	var sct core.SignedCertificateTimestamp
+	for !done && retries <= pub.CT.SubmissionRetries {
+		resp, err := postJSON(&client, fmt.Sprintf("%s%s", log.URI, "/ct/v1/add-chain"), jsonSubmission, &sct)
+		if err != nil {
+			// Retry the request, log the error
+			// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+			pub.log.AuditErr(fmt.Errorf("Error POSTing JSON to CT log submission endpoint [%s]: %s", log.URI, err))
+			if retries >= pub.CT.SubmissionRetries {
+				break
+			}
+			retries++
+			time.Sleep(pub.CT.SubmissionBackoff)
+			continue
+		} else {
+			if resp.StatusCode == http.StatusRequestTimeout || resp.StatusCode == http.StatusServiceUnavailable {
+				// Retry the request after either 10 seconds or the period specified
+				// by the Retry-After header
+				backoff := pub.CT.SubmissionBackoff
+				if retryAfter := resp.Header.Get("Retry-After"); retryAfter != "" {
+					if seconds, err := strconv.Atoi(retryAfter); err != nil {
+						backoff = time.Second * time.Duration(seconds)
+					}
+				}
+				if retries >= pub.CT.SubmissionRetries {
+					break
+				}
+				retries++
+				time.Sleep(backoff)
+				continue
+			} else if resp.StatusCode != http.StatusOK {
+				// Not something we expect to happen, set error, break loop and log
+				// the error
+				// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+				pub.log.AuditErr(fmt.Errorf("Unexpected status code returned from CT log submission endpoint [%s]: Unexpected status code [%d]", log.URI, resp.StatusCode))
+				break
+			}
+		}
+
+		done = true
+		break
+	}
+	if !done {
+		pub.log.Warning(fmt.Sprintf(
+			"Unable to submit certificate to CT log [Serial: %s, Log URI: %s, Retries: %d]",
+			serial,
+			log.URI,
+			retries,
+		))
+		return fmt.Errorf("Unable to submit certificate")
+	}
+
+	if err := sct.CheckSignature(); err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(err)
+		return err
+	}
+
+	pub.log.Notice(fmt.Sprintf(
+		"Submitted certificate to CT log [Serial: %s, Log URI: %s, Retries: %d, Signature: %x]",
+		serial,
+		log.URI,
+		retries, sct.Signature,
+	))
+
+	// Set certificate serial and add SCT to DB
+	sct.CertificateSerial = serial
+	err := pub.SA.AddSCTReceipt(sct)
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), "duplicate entry") {
```

Comment:
This is an error-prone way to check for specific error types. I'm assuming gorp doesn't give nice error types? If not, we should convert this to a specific error type within SA and check for it here.